pkg/virt-controller/watch/vm.go

```diff
@@ -640,11 +642,39 @@ func (c *VMController) updatePVCMemoryDumpAnnotation(vm *virtv1.VirtualMachine)
 }
 
 func (c *VMController) VMICPUsPatch(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
-	test := fmt.Sprintf(`{ "op": "test", "path": "/spec/domain/cpu/sockets", "value": %s}`, strconv.FormatUint(uint64(vmi.Spec.Domain.CPU.Sockets), 10))
-	update := fmt.Sprintf(`{ "op": "replace", "path": "/spec/domain/cpu/sockets", "value": %s}`, strconv.FormatUint(uint64(vm.Spec.Template.Spec.Domain.CPU.Sockets), 10))
-	patch := fmt.Sprintf("[%s, %s]", test, update)
+	patchSet := patch.New(
+		patch.WithTest("/spec/domain/cpu/sockets", vmi.Spec.Domain.CPU.Sockets),
+		patch.WithReplace("/spec/domain/cpu/sockets", vm.Spec.Template.Spec.Domain.CPU.Sockets),
+	)
+
+	vcpusDelta := hardware.GetNumberOfVCPUs(vm.Spec.Template.Spec.Domain.CPU) - hardware.GetNumberOfVCPUs(vmi.Spec.Domain.CPU)
+	resourcesDelta := resource.NewMilliQuantity(vcpusDelta*int64(1000/c.clusterConfig.GetCPUAllocationRatio()), resource.DecimalSI)
 
-	_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, []byte(patch), metav1.PatchOptions{})
+	if !vm.Spec.Template.Spec.Domain.Resources.Requests.Cpu().IsZero() {
+		newCpuReq := vm.Spec.Template.Spec.Domain.Resources.Requests.Cpu().DeepCopy()
+		newCpuReq.Add(*resourcesDelta)
+
+		patchSet.AddOption(
+			patch.WithTest("/spec/domain/resources/requests/cpu", vmi.Spec.Domain.Resources.Requests.Cpu().String()),
+			patch.WithReplace("/spec/domain/resources/requests/cpu", newCpuReq.String()),
+		)
+	}
+	if !vm.Spec.Template.Spec.Domain.Resources.Limits.Cpu().IsZero() {
+		newCpuLimit := vm.Spec.Template.Spec.Domain.Resources.Limits.Cpu().DeepCopy()
+		newCpuLimit.Add(*resourcesDelta)
+
+		patchSet.AddOption(
+			patch.WithTest("/spec/domain/resources/limits/cpu", vmi.Spec.Domain.Resources.Limits.Cpu().String()),
+			patch.WithReplace("/spec/domain/resources/limits/cpu", newCpuLimit.String()),
+		)
+	}
```

Comment:
Could we log a message saying we're altering requests/limits?

Reply:
done!

Reply:
@acardace @jean-edouard Hey! can you please elaborate on the need to log this event? is this something unexpected to the user that initiates the change?

Reply:
IMO, cases where we alter values that were explicitly set by the user should definitely trigger at least a print.